pkg/domain/infra/tunnel/containers.go

```diff
@@ -507,33 +509,83 @@ func (ic *ContainerEngine) ContainerRun(ctx context.Context, opts entities.Conta
 	for _, w := range con.Warnings {
 		fmt.Fprintf(os.Stderr, "%s\n", w)
 	}
-	report := entities.ContainerRunReport{Id: con.ID}
-	// Attach
-	if !opts.Detach {
-		err = startAndAttach(ic, con.ID, &opts.DetachKeys, opts.InputStream, opts.OutputStream, opts.ErrorStream)
-		if err == nil {
-			exitCode, err := containers.Wait(ic.ClientCxt, con.ID, nil)
-			if err == nil {
-				report.ExitCode = int(exitCode)
+
+	if opts.Rm {
+		// Defer the removal, so we can return early if needed and
+		// de-spaghetti the code.
+		defer func() {
+			if err := containers.Remove(ic.ClientCxt, con.ID, bindings.PFalse, bindings.PTrue); err != nil {
```

Comment:
I think we don't want to do this in the opts.Detach case - for that, we rely entirely on the cleanup process. Should be enough to move this block down to 536

Reply:
Good point. The if start-attach failed, the error should also go on debug (as done locally).